util/buildinfo/buildinfo.go

```diff
@@ -2,136 +2,208 @@ package buildinfo
 
 import (
 	"context"
+	"encoding/base64"
 	"encoding/json"
 	"sort"
+	"strings"
 
 	"github.com/docker/distribution/reference"
-	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	"github.com/moby/buildkit/frontend"
 	"github.com/moby/buildkit/source"
+	binfotypes "github.com/moby/buildkit/util/buildinfo/types"
 	"github.com/moby/buildkit/util/urlutil"
 	"github.com/pkg/errors"
 )
 
-const ImageConfigField = "moby.buildkit.buildinfo.v1"
+// Decode decodes a base64 encoded build info.
+func Decode(enc string) (bi binfotypes.BuildInfo, _ error) {
+	dec, err := base64.StdEncoding.DecodeString(enc)
+	if err != nil {
+		return bi, err
+	}
+	err = json.Unmarshal(dec, &bi)
+	return bi, err
+}
 
-// Merge combines and fixes build info from image config
-// key moby.buildkit.buildinfo.v1.
-func Merge(ctx context.Context, buildInfo map[string]string, imageConfig []byte) ([]byte, error) {
-	icbi, err := imageConfigBuildInfo(imageConfig)
+// Get returns build info.
+func Get(ctx context.Context, req frontend.SolveRequest, buildSources map[string]string, imageConfig []byte) ([]byte, error) {
+	icbi, err := FromImageConfig(imageConfig)
 	if err != nil {
 		return nil, err
 	}
+	srcs, err := mergeSources(ctx, buildSources, icbi)
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(binfotypes.BuildInfo{
+		Frontend: req.Frontend,
+		Attrs:    filterAttrs(req.FrontendOpt),
+		Sources:  srcs,
+	})
+}
 
+// StripAttrs removes or not build attributes from build info.
+func StripAttrs(dt []byte, keep bool) (_ []byte, err error) {
```

Comment:
nit: I think if this is called `StripAttrs` then it should not have a second parameter as the name implies it. With the second parameter it could maybe be `Format(dt, FormatOpts{Attrs: false})`.